doc/debug_framework_design_spec.md

```diff
@@ -0,0 +1,394 @@
+| Post-action                   |  "upload" / "compress-rotate-keep"  | compress-rotate-keep          |
+| Server-location               |  ipaddress                          | 127.0.0.1                     |
+| Upload-method                 |  "scp" / "tftp"                     | tftp                          |
+| Upload-directory              |  "dir_path" / "default_dir"         | default_dir                   | 
+
+
+### 3.1.2 Assert Framework
+
+#### 3.1.2.1 Overview
+Asserts are added in the program execution sequence to confirm that the data/state at a certain point is valid/true. During developement, if the programming sequence fails in an assert condition then the program execution is stopped by crash/exception. In production code, asserts are normally removed. This framework enhances/extendes the assert to provide more debug details when an assert fails.  
+
+Classify assert failure conditions based on following types, assert() will have type and the module as additional arguments  
+- DUMP: Invokes the debug framework registered callback API corresponding to the module
+- BTRACE: Prints bracktrace and continue
+- SYSLOG: Update syslog with the assert failure
+- ABORT: Stop/throw exception
+
+
+#### 3.1.2.2 PsuedoCode:
+
+```
+    static void custom_assert(bool exp, const char*func, const unsigned line);  
+
+    #ifdef assert
+    #undef assert
+    #define assert(exp) Debugframework::custom_assert(exp, __PRETTY_FUNCTION__, __LINE__)
+    #endif
+```
+
+## 3.2 DB Changes
+
+### 3.2.1 APP DB
+
+**Dump table**
+
+For triggering dump routines  
+
+```
+; Defines schema for DEBUG Event attributes
+key               = DAEMON:daemon_name              ; daemon_session_name is unique identifier;  
+; field           = value  
+DUMP_TYPE         = "short" / "full"                ; summary dump or full dump  
+DUMP_TARGET       = "default" / "syslog"            ; syslog or specified file  
+PASSTHRU_ARGS     = arglist                         ; zero or more strings separated by ","
+
+```
+**Dump done table**
+```
+; Defines schema for DEBUG response attributes
+key                = DAEMON:daemon_name             ; daemon_session_name is unique identifier;
+; field            = value  
+RESPONSE_TYPE      = < enum DumpResponse::result >  ; pending/failure/successful  
+```
+
+## 3.3 CLI
+
+### 3.3.1 Show Commands
+
+|Syntax                         | Description                                                                   |
+|-------------------------------|-------------------------------------------------------------------------------|
+|show debug all                 | This command will invoke dump routines for all components with default action |
+|show debug < component >       | This command will invoke dump routine for specific component                  |
+|show debug actions < options > | This command will display the configured framework actions                    |  

```

Comment:
how do we know all the component that has such debug capability? is there a command for that?

Reply:
component owner trying to debug his component will enter component name so that the trigger to dump will be specific to the component specified.  If the admin is trying to collect debug information and has no knowledge of registered components he may consider show debug all.   

If its an engineer but not component owner who is trying to figure out the registered components he can query Redis DB for the registered component list using redis-cli and then proceed with the trigger.